### PR TITLE
OCPBUGS-29596: e2e: Enhance tests related to crio annotations

### PR DESCRIPTION
**Pull request by mrniranjan**
1. Add new test case  for testing cpu-quota.crio.io annotation
2. Move the existing cpu load balancing test under single container where all the tests related to crio annotation exist
3. Modify tests to be executed under different runtimes and cgroup versions

### CODE REVIEW

File: test/e2e/performanceprofile/functests/utils/cgroup/controller/controller.go [hunk 1]
```diff
@@ -14,7 +14,10 @@ type CpuSet struct {
...
 
 type Cpu struct {
-	Quota   string
-	Period  string
-	cpuStat map[string]string
+	Quota  string
+	Period string
+}
+
+type CpuStat struct {
```
Comment (Tal-or) [inline test/e2e/performanceprofile/functests/utils/cgroup/controller/controller.go:21]: Each struct should represent a cgroup controller. i.e `cpu`,`cpuset`,`memory`,`pids`...
`stats` is not a controller, it's a field under the `cpu` controller
Could you please elaborate why this change is needed?

File: test/e2e/performanceprofile/functests/utils/cgroup/v2/v2.go [hunk 2]
```diff
@@ -60,6 +60,32 @@ func (cm *ControllersManager) Cpu(ctx context.Context, pod *corev1.Pod, containe
...
+	}
+	b, err := pods.ExecCommandOnPod(cm.k8sClient, pod, containerName, cmd)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve cgroup config for pod. pod=%q, container=%q; %w", client.ObjectKeyFromObject(pod).String(), containerName, err)
+	}
+	output := strings.TrimSpace(string(b))
+	interfacevalues := strings.Split(output, "\r\n")
+	// cpu.stat always contains 3 stats usage_usec, user_usec, system_usec
+	// only when cpu controller is enabled other stats like nr_periods etc are enabled
+	if len(interfacevalues) < 4 {
```
Comment (Tal-or) [inline test/e2e/performanceprofile/functests/utils/cgroup/v2/v2.go:79]: A more robust way to check whether cpu controller enabled on cgroup v2 is by checking the `cgroup.controllers` and `cgroup.subtree_control` files.
From the documentation https://man7.org/linux/man-pages/man7/cgroups.7.html:
_Active cgroups must be specified via the files
cgroup.controllers and cgroup.subtree_control._

Comment (Tal-or) [reply test/e2e/performanceprofile/functests/utils/cgroup/v2/v2.go:79]: It's only a suggestion, if we want to keep the function identical this is also an option but I think it would be better to move the common (identical for v1 and v2) functions into a separate package.
This can be handled in a separate PR as well

File: test/e2e/performanceprofile/functests/1_performance/cpu_management.go [hunk 3]
```diff
@@ -654,6 +529,126 @@ var _ = Describe("[rfe_id:27363][performance] CPU Management", Ordered, func() {
...
+				err := getter.Container(ctx, testpod, testpod.Spec.Containers[0].Name, cpusetCfg)
+				Expect(err).ToNot(HaveOccurred())
+				// Get cpus used by the container
+				tasksetcmd := []string{"taskset", "-pc", "1"}
+				testpodAffinity, err := pods.ExecCommandOnPod(testclient.K8sClient, testpod, testpod.Spec.Containers[0].Name, tasksetcmd)
+				podCpusStr := string(testpodAffinity)
+				parts := strings.Split(strings.TrimSpace(podCpusStr), ":")
+				testpodCpus := strings.TrimSpace(parts[1])
+				testlog.Infof("%v pod is using %v cpus", testpod.Name, string(testpodCpus))
+				Expect(cpusetCfg.Cpus).To(Equal(testpodCpus), "cpuset.cpus not matching the process affinity")
```
Comment (Tal-or) [inline test/e2e/performanceprofile/functests/1_performance/cpu_management.go:610]: The comparison here is done between strings but they can be logically equals while unequal by string.
Example:
setA: `1-3,7`
setB: `1,2,3,7`
logically equals but the strings are different.
parse them into `cpuset` and then compare.

File: test/e2e/performanceprofile/functests/utils/cgroup/controller/controller.go (hunk 1 above)
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/utils/cgroup/controller/controller.go:21]: We discussed this before stat has multi line values where the number files are different in v1 and v2 and with current implementation reading stat is difficult where the current implementation tries to read multiple files and thats a problem.

File: test/e2e/performanceprofile/functests/1_performance/cpu_management.go [hunk 4]
```diff
@@ -654,6 +529,126 @@ var _ = Describe("[rfe_id:27363][performance] CPU Management", Ordered, func() {
...
+			time.Sleep(30 * time.Second) // let cpumanager reconcile loop catch up
+
+			// It's possible that when this test runs the value of
+			// defaultCpuNotInSchedulingDomains is empty if no gu pods are running
+			defaultCpuNotInSchedulingDomains, err := getCPUswithLoadBalanceDisabled(ctx, workerRTNode)
+			Expect(err).ToNot(HaveOccurred(), "Unable to fetch scheduling domains")
+			if len(defaultCpuNotInSchedulingDomains) > 0 {
+				pods, err := pods.GetPodsOnNode(context.TODO(), workerRTNode.Name)
+				if err != nil {
+					testlog.Warningf("cannot list pods on %q: %v", workerRTNode.Name, err)
```
[the remark this answers is not in the document]
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/1_performance/cpu_management.go:561]: hmm, i am not sure too, though this code is mostly used as debug to know that we found some pods before the test starts and for some reason we are unable to list the pods before throwing the error: " the test expects all CPUs within a scheduling domain when starting"

File: test/e2e/performanceprofile/functests/utils/cgroup/v2/v2.go (hunk 2 above)
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/utils/cgroup/v2/v2.go:79]: Yeah i will address this in separate PR , for now i will modify the message.

File: test/e2e/performanceprofile/functests/1_performance/cpu_management.go [hunk 5]
```diff
@@ -507,11 +381,12 @@ var _ = Describe("[rfe_id:27363][performance] CPU Management", Ordered, func() {
 
 			containersCgroups = strings.Trim(containersCgroups, "\n")
 			containersCgroupsDirs := strings.Split(containersCgroups, "\n")
-			Expect(len(containersCgroupsDirs)).To(Equal(2), "unexpected amount of containers cgroups")
+			//Expect(len(containersCgroupsDirs)).To(Equal(2), "unexpected amount of containers cgroups")
```
[the remark this answers is not in the document]
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/1_performance/cpu_management.go:384]: fixed in the latest commit

File: test/e2e/performanceprofile/functests/1_performance/cpu_management.go [hunk 6]
```diff
@@ -654,6 +529,126 @@ var _ = Describe("[rfe_id:27363][performance] CPU Management", Ordered, func() {
 		)
 
 	})
+	When("Crio Annotations", func() {
```
[the remark this answers is not in the document]
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/1_performance/cpu_management.go]: fixed in the latest commit

File: test/e2e/performanceprofile/functests/1_performance/cpu_management.go [hunk 7]
```diff
@@ -654,6 +529,126 @@ var _ = Describe("[rfe_id:27363][performance] CPU Management", Ordered, func() {
...
+			Expect(err).ToNot(HaveOccurred(), "failed to create guaranteed pod %v", testpod)
+			allTestpods[testpod.UID] = testpod
+		})
+
+		AfterAll(func() {
+			for podUID, testpod := range allTestpods {
+				testlog.Infof("deleting test pod %s/%s UID=%q", testpod.Namespace, testpod.Name, podUID)
+				err := testclient.Client.Get(ctx, client.ObjectKeyFromObject(testpod), testpod)
+				Expect(err).ToNot(HaveOccurred())
+				//testpodUID := testpod.UID
```
[the remark this answers is not in the document]
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/1_performance/cpu_management.go]: fixed in the latest commit

File: test/e2e/performanceprofile/functests/1_performance/cpu_management.go (hunk 3 above)
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/1_performance/cpu_management.go]: fixed in the latest commit

File: test/e2e/performanceprofile/functests/utils/cgroup/controller/controller.go (hunk 1 above)
Comment (Tal-or) [reply test/e2e/performanceprofile/functests/utils/cgroup/controller/controller.go:21]: @mrniranjan Can't you simply read the `cpu.stat` in a separate call and store it in `cfg.cpuStat`?
Add the logic here:
https://github.com/openshift/cluster-node-tuning-operator/blob/master/test/e2e/performanceprofile/functests/utils/cgroup/v1/v1.go#L45 
And here:
https://github.com/openshift/cluster-node-tuning-operator/blob/master/test/e2e/performanceprofile/functests/utils/cgroup/v2/v2.go#L45

something like:
```
cmd = []string{
		"/bin/cat",
		dirPath + "/cpu.stat",
	}
	b, err := pods.ExecCommandOnPod(cm.k8sClient, pod, containerName, cmd)
// do some parsing logic and store in map
```

Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/utils/cgroup/controller/controller.go]: addressed in the latest commit

File: test/e2e/performanceprofile/functests/1_performance/cpu_management.go [hunk 8]
```diff
@@ -479,14 +348,19 @@ var _ = Describe("[rfe_id:27363][performance] CPU Management", Ordered, func() {
...
+			var cpusetPath string
 			// find used because that crictl does not show infra containers, `runc list` shows them
 			// but you will need somehow to find infra containers ID's
 			podUID := strings.Replace(string(testpod.UID), "-", "_", -1)
-
 			podCgroup := ""
+			if cgroupV2 {
+				cpusetPath = "/rootfs/sys/fs/cgroup/kubepods.slice"
+			} else {
+				cpusetPath = "/rootfs/sys/fs/cgroup/cpuset"
```
Comment (Tal-or) [inline test/e2e/performanceprofile/functests/1_performance/cpu_management.go:359]: On v1 the `kubepods.slice` part exists as well

Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/1_performance/cpu_management.go:359]: correct, on v1 the paths differ

File: test/e2e/performanceprofile/functests/utils/cgroup/v2/v2.go [hunk 9]
```diff
@@ -77,7 +77,7 @@ func (cm *ControllersManager) CpuStat(ctx context.Context, pod *corev1.Pod, cont
 	// cpu.stat always contains 3 stats usage_usec, user_usec, system_usec
 	// only when cpu controller is enabled other stats like nr_periods etc are enabled
 	if len(interfacevalues) < 4 {
-		return nil, fmt.Errorf("CPU Controller is not enabled")
+		return nil, fmt.Errorf("cpu.stat doesn't contain all the cpu metrics required")
```
Comment (shajmakh) [inline test/e2e/performanceprofile/functests/utils/cgroup/v2/v2.go]: minor: this was restored in the later commit:
https://github.com/openshift/cluster-node-tuning-operator/pull/955/commits/8092af08081d8334505aaa3bc148f2c221a24cba#diff-f2536864a27be1f85973bce12efd1a1447d9fab4bb03170997c538c5e5650e81R81
if not intended to change it here, I think better to discard this line change

File: test/e2e/performanceprofile/functests/1_performance/cpu_management.go [hunk 10]
```diff
@@ -596,23 +595,27 @@ var _ = Describe("[rfe_id:27363][performance] CPU Management", Ordered, func() {
...
 				podCpusStr := string(testpodAffinity)
 				parts := strings.Split(strings.TrimSpace(podCpusStr), ":")
 				testpodCpus := strings.TrimSpace(parts[1])
-				testlog.Infof("%v pod is using %v cpus", testpod.Name, string(testpodCpus))
-				Expect(cpusetCfg.Cpus).To(Equal(testpodCpus), "cpuset.cpus not matching the process affinity")
+				testlog.Infof("%v pod is using %v cpus", testpod.Name, testpodCpus)
+				podAffinityCpuset, err := cpuset.Parse(testpodCpus)
+				Expect(err).ToNot(HaveOccurred(), "Unable to parse cpus %s used by %s pod", testpodCpus, testpod.Name)
+				cgroupCpuset, err := cpuset.Parse(cpusetCfg.Cpus)
+				Expect(err).ToNot(HaveOccurred(), "Unable to parse cpus from cgroups.cpugset")
```
[the remark this answers is not in the document]
Comment (Tal-or) [reply test/e2e/performanceprofile/functests/1_performance/cpu_management.go:612]: probably typo

File: test/e2e/performanceprofile/functests/1_performance/cpu_management.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Tal-or) [reply test/e2e/performanceprofile/functests/1_performance/cpu_management.go:384]: The comment is still here, probably some rebase issue

File: test/e2e/performanceprofile/functests/utils/cgroup/v1/v1.go [hunk 11]
```diff
@@ -57,9 +57,36 @@ func (cm *ControllersManager) Cpu(ctx context.Context, pod *corev1.Pod, containe
 	output := strings.Split(string(b), "\r\n")
 	cfg.Quota = output[0]
 	cfg.Period = output[1]
+	cfg.Stat, err = stat(cm.k8sClient, pod, containerName, childName)
```
Comment (Tal-or) [inline test/e2e/performanceprofile/functests/utils/cgroup/v1/v1.go:60]: We should check the `error` value

Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/utils/cgroup/v1/v1.go:60]: fixed in the latest commit

File: test/e2e/performanceprofile/functests/utils/cgroup/v2/v2.go (hunk 9 above)
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/utils/cgroup/v2/v2.go]: fixed in the latest commit

File: test/e2e/performanceprofile/functests/1_performance/cpu_management.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/1_performance/cpu_management.go]: fixed in the latest commit

File: test/e2e/performanceprofile/functests/1_performance/cpu_management.go [hunk 12]
```diff
@@ -654,6 +529,126 @@ var _ = Describe("[rfe_id:27363][performance] CPU Management", Ordered, func() {
...
+			}
+		})
+
+		Describe("cpuset controller", func() {
+			It("Verify cpuset controller interface files have right values", func() {
+				cpusetCfg := &controller.CpuSet{}
+				err := getter.Container(ctx, testpod, testpod.Spec.Containers[0].Name, cpusetCfg)
+				Expect(err).ToNot(HaveOccurred())
+				// Get cpus used by the container
+				tasksetcmd := []string{"taskset", "-pc", "1"}
```
[the remark this answers is not in the document]
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/1_performance/cpu_management.go]: fixed in the latest commit

File: test/e2e/performanceprofile/functests/1_performance/cpu_management.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/1_performance/cpu_management.go]: fixed in the latest commit

File: test/e2e/performanceprofile/functests/1_performance/cpu_management.go [hunk 13]
```diff
@@ -654,6 +528,129 @@ var _ = Describe("[rfe_id:27363][performance] CPU Management", Ordered, func() {
...
+						testlog.Infof("- %s/%s %s", pod.Namespace, pod.Name, pod.UID)
+					}
+					testlog.Infof("pods on %q END", workerRTNode.Name)
+				}
+				Expect(defaultCpuNotInSchedulingDomains).To(BeEmpty(), "the test expects all CPUs within a scheduling domain when starting")
+			}
+
+			By("Starting the pod")
+			testpod.Spec.NodeSelector = testutils.NodeSelectorLabels
+			testpod.Spec.Containers[0].Image = "quay.io/mniranja/busycpus"
```
Comment (yanirq) [inline test/e2e/performanceprofile/functests/1_performance/cpu_management.go:573]: we should probably replace this quay url

Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/1_performance/cpu_management.go]: @yanirq fixed it, using http://quay.io/ocp-edge-qe/busycpus

**Comment by yanirq** [comment]
@mrniranjan  does this PR warrant all the commits ? (9 commits)   perhaps we should squash them

**Comment by mrniranjan** [comment]
@yanirq i have squashed the commits
